### DISCUSSION TITLE
Can I make eval use custom functions?

### DISCUSSION THREAD

**Discussion by onomou** [typst/typst#3782] [General]
I am creating an exam using a question bank, and want to use custom functions defined in my Typst file. The questions are loaded with `
#let questions = json("questions.json")` and rendered with `#eval(question.at("questiontext", mode: "markup")`. This works fine if I use only built-in functions: 

`main.typ`
```typst
#let questions = json("questions.json")
#for question in questions [
    + #eval(question.text, mode: "markup")
]
```

`questions.json`
```json
[
  {
      "text": "Graph the function $f(x)=x^2+2x+1$."
  },
  {
    "text": "#lorem(3)"
  },
  {
    "text": "Complete the square: $x^2-6 x + #box(width: 16pt, height: 16pt, stroke: 0.5pt + black, baseline: 4pt) = (#box(width: 36pt))^2$."
  }
]
```

It works if the function is defined in the string fed to `eval`:

```typst
#let fulldef = "#let myfunction = ([This is a useful function.])
#myfunction"

#eval(mode: "markup", fulldef) // this works, since myfunction is defined within the eval
```

I can't make it work with custom functions though. Here is a MWE:
```typst
#let thedefinition = "#let myfunction = ()"
#let theevaluation = "#myfunction"

#eval(mode: "markup", thedefinition)
#eval(mode: "markup", theevaluation) // Error: Unknown variable: myfunction
```

Any workarounds, or alternatives to achieve my goal?

https://typst.app/project/rOJBSBL4mjggccIIPoDOCE

**Comment by ludwig-austermann** [comment]
Take a look at the `scope` argument of the `eval` function. You can pass your function that way.

**Comment by onomou** [comment]
That works -- thank you! I can pass functions defined in my Typst file to `eval` to use with my JSON now:
```typst
#let thefunction = [Local function]
#eval("#thefunction", mode: "markup", scope: (thefunction: thefunction))
```
